perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -2578,4 +2585,64 @@ private boolean checkGroupsCycle(PerunSession sess, int resultGroupId, int opera
 
 		return false;
 	}
+
+	/**
+	 * Put Group to the queue defined as groups to be synchronized as soon as possible (FIFO)
+	 * One exception, if putAsFirst is set to true, this group will skip the order and will be added as first to
+	 * process (LIFO).
+	 *
+	 * WARNING:
+	 * - using putAsFirst set to true can in specific situation (group is already on first place in queue)
+	 * delayed processing of group
+	 * - method is synchronized so only one thread should add group to the queue at one time (this prevents
+	 * creating duplicities of groups in queue), but it also lasts more time if call at bulk
+	 *
+	 * @param group group to add to the queue
+	 * @param putAsFirst group will be put to the first place in queue (LIFO)
+	 * @return true if group was added, false if group was already in queue on any place, it always return true for synchronizeAsFirst
+	 * @throws InternalErrorException if group is null
+	 */
+	private synchronized boolean putGroupToQueueToBeSynchronized(Group group, boolean putAsFirst) throws InternalErrorException {
```

Comment:
The method `putGroupToQueueToBeSynchronized()` should not have the `synchronized` modifier, which synchronizes on "this" object, but it should rather have its content wrapped in `synchronized(queueOfGroupsToBeSynchronized){ ...}` which synchronizes on the queue itself. Otherwise the test for contains and addlast may not prevent duplicities because situatin may change between the test and the call.